In [2]:
USE AdventureWorks2019
GO

Commands completed successfully.

Total execution time: 00:00:00.191

Q1. How many salespersons has bonus less than 5,000.00 Sales.SalesPerson?

In [1]:
SELECT COUNT(BusinessEntityID) AS Total_Salespersons 
FROM Sales.SalesPerson
WHERE Bonus < 5000.00

(1 row affected)

Total execution time: 00:00:00.196

Total_Salespersons 12

Q2. How many salespersons in the list of TerritoryID equals to 4?

In [3]:
SELECT COUNT(BusinessEntityID) AS Total_Salespersons 
FROM Sales.SalesPerson
WHERE TerritoryID = 4

(1 row affected)

Total execution time: 00:00:00.198

Total_Salespersons 2

Q3. How many salespersons records that have been modified on 2011-05-24?

In [4]:
SELECT COUNT(BusinessEntityID) AS Total_Salespersons 
FROM Sales.SalesPerson
WHERE ModifiedDate = '2011-05-24'

(1 row affected)

Total execution time: 00:00:00.201

Total_Salespersons 9

Q4. On product table (Production.Product), create an analytic query against it by color with the following statistics functions

 <span style="background-color: rgba(127, 127, 127, 0.1);">Sum, Count, Average, Minimum, Maximum and Standard Deviation. Also, filter the non-color products out of this list.</span>

<span style="background-color: rgba(127, 127, 127, 0.1);">Note: Find the aggregation from StandardCost column.</span>

In [11]:
SELECT Color,
       SUM(StandardCost) AS Sum,
       COUNT(StandardCost) AS Count,
       AVG(StandardCost) AS Average,
       MIN(StandardCost) AS Minimum,
       MAX(StandardCost) AS Maximum,
       STDEV(StandardCost) AS Standard_Deviation
FROM Production.Product
WHERE Color IS NOT NULL
GROUP BY Color

(9 rows affected)

Total execution time: 00:00:00.213

Color Sum Count Average Minimum Maximum Standard_Deviation Black 38636.5002 93 415.4462 0.0000 1898.0944 510.2448347755234 Blue 14746.1464 26 567.1594 13.0863 1481.9379 465.4786631864363 Grey 51.5625 1 51.5625 51.5625 51.5625 NULL Multi 272.2542 8 34.0317 6.9223 38.4923 10.974885926351536 Red 32610.7661 38 858.1780 13.0863 2171.2942 657.2815651545327 Silver 20060.0483 43 466.5127 0.0000 1912.1544 581.4484410160618 Silver/Black 198.9700 7 28.4242 17.9776 35.9596 8.063041397104998 White 13.5172 4 3.3793 3.3623 3.3963 0.019629909152522135 Yellow 21507.6521 36 597.4347 41.5723 1481.9379 432.1348945862517

Q.4.1. How can we find the product color with the average that less than or equal to 500?

In [14]:
SELECT Color
FROM Production.Product
WHERE Color IS NOT NULL
GROUP BY Color
HAVING AVG(StandardCost) <= 500

(6 rows affected)

Total execution time: 00:00:00.211

Color Black Grey Multi Silver Silver/Black White

Q5. Create a SQL script that list the employee in the following order of columns: JobTitle, FirstName and LastName.

Join Hint: HumanResources.Employee and Person.Person

In [15]:
SELECT JobTitle, FirstName, LastName
FROM HumanResources.Employee
JOIN Person.Person ON HumanResources.Employee.BusinessEntityID = Person.Person.BusinessEntityID

(290 rows affected)

Total execution time: 00:00:00.212

JobTitle FirstName LastName Chief Executive Officer Ken Sánchez Vice President of Engineering Terri Duffy Engineering Manager Roberto Tamburello Senior Tool Designer Rob Walters Design Engineer Gail Erickson Design Engineer Jossef Goldberg Research and Development Manager Dylan Miller Research and Development Engineer Diane Margheim Research and Development Engineer Gigi Matthew Research and Development Manager Michael Raheem Senior Tool Designer Ovidiu Cracium Tool Designer Thierry D'Hers Tool Designer Janice Galvin Senior Design Engineer Michael Sullivan Design Engineer Sharon Salavaria Marketing Manager David Bradley Marketing Assistant Kevin Brown Marketing Specialist John Wood Marketing Assistant Mary Dempsey Marketing Assistant Wanida Benshoof Marketing Specialist Terry Eminhizer Marketing Specialist Sariya Harnpadoungsataya Marketing Specialist Mary Gibson Marketing Specialist Jill Williams Vice President of Production James Hamilton Production Control Manager Peter Krebs Production Supervisor - WC60 Jo Brown Production Technician - WC60 Guy Gilbert Production Technician - WC60 Mark McArthur Production Technician - WC60 Britta Simon Production Technician - WC60 Margie Shoop Production Technician - WC60 Rebecca Laszlo Production Technician - WC60 Annik Stahl Production Technician - WC60 Suchitra Mohan Production Technician - WC60 Brandon Heidepriem Production Technician - WC60 Jose Lugo Production Technician - WC60 Chris Okelberry Production Technician - WC60 Kim Abercrombie Production Technician - WC60 Ed Dudenhoefer Production Supervisor - WC60 JoLynn Dobney Production Technician - WC60 Bryan Baker Production Technician - WC60 James Kramer Production Technician - WC60 Nancy Anderson Production Technician - WC60 Simon Rapier Production Technician - WC60 Thomas Michaels Production Technician - WC60 Eugene Kogan Production Supervisor - WC10 Andrew Hill Production Technician - WC10 Ruth Ellerbrock Production Technician - WC10 Barry Johnson Production Technician - WC10 Sidney Higa Production Technician - WC10 Jeffrey Ford Production Technician - WC10 Doris Hartwig Production Technician - WC10 Diane Glimp Production Technician - WC10 Bonnie Kearney Production Supervisor - WC50 Taylor Maxwell Production Technician - WC50 Denise Smith Production Technician - WC50 Frank Miller Production Technician - WC50 Kendall Keil Production Technician - WC50 Bob Hohman Production Technician - WC50 Pete Male Production Technician - WC50 Diane Tibbott Production Supervisor - WC60 John Campbell Production Technician - WC60 Maciej Dusza Production Technician - WC60 Michael Zwilling Production Technician - WC60 Randy Reeves Production Technician - WC60 Karan Khanna Production Technician - WC60 Jay Adams Production Technician - WC60 Charles Fitzgerald Production Technician - WC60 Steve Masters Production Technician - WC60 David Ortiz Production Supervisor - WC30 Michael Ray Production Technician - WC30 Steven Selikoff Production Technician - WC30 Carole Poland Production Technician - WC30 Bjorn Rettig Production Technician - WC30 Michiko Osada Production Technician - WC30 Carol Philips Production Technician - WC30 Merav Netz Production Supervisor - WC40 Reuben D'sa Production Technician - WC40 Eric Brown Production Technician - WC40 Sandeep Kaliyath Production Technician - WC40 Mihail Frintu Production Technician - WC40 Jack Creasey Production Technician - WC40 Patrick Cook Production Technician - WC40 Frank Martinez Production Technician - WC40 Brian Goldstein Production Technician - WC40 Ryan Cornelsen Production Supervisor - WC10 Cristian Petculescu Production Technician - WC10 Betsy Stadick Production Technician - WC10 Patrick Wedge Production Technician - WC10 Danielle Tiedt Production Technician - WC10 Kimberly Zimmerman Production Technician - WC10 Tom Vande Velde Production Supervisor - WC50 Kok-Ho Loh Production Technician - WC50 Russell Hunter Production Technician - WC50 Jim Scardelis Production Technician - WC50 Elizabeth Keyser Prod

6\. 2 tables contain the information of vendors (Purchasing.Vendor) and vendor’s product (Purchasing.ProductVendor)

6.1. How many rows that vendor contains no products?

In [16]:
SELECT BusinessEntityID FROM Purchasing.Vendor
UNION 
SELECT BusinessEntityID FROM Purchasing.ProductVendor
ORDER BY BusinessEntityID

(104 rows affected)

Total execution time: 00:00:00.211

BusinessEntityID 1492 1494 1496 1498 1500 1502 1504 1506 1508 1510 1512 1514 1516 1518 1520 1522 1524 1526 1528 1530 1532 1534 1536 1538 1540 1542 1544 1546 1548 1550 1552 1554 1556 1558 1560 1562 1564 1566 1568 1570 1572 1574 1576 1578 1580 1582 1584 1586 1588 1590 1592 1594 1596 1598 1600 1602 1604 1606 1608 1610 1612 1614 1616 1618 1620 1622 1624 1626 1628 1630 1632 1634 1636 1638 1640 1642 1644 1646 1648 1650 1652 1654 1656 1658 1660 1662 1664 1666 1668 1670 1672 1674 1676 1678 1680 1682 1684 1686 1688 1690 1692 1694 1696 1698

In [17]:
SELECT COUNT(Purchasing.Vendor.BusinessEntityID) AS Total_Vendor
FROM Purchasing.Vendor
JOIN Purchasing.ProductVendor ON Purchasing.ProductVendor.BusinessEntityID = Purchasing.Vendor.BusinessEntityID
WHERE Purchasing.ProductVendor.ProductID IS NULL

(1 row affected)

Total execution time: 00:00:00.194

Total_Vendor 0

6.2. What methodology to achieve that?

**Answer:** 

I used 2 methods, firstly I check vendor rows it has 104 rows and then I made a union check between 2 table because Purchasing.ProductVendor it contains only 

the vendor that have a product based on ProductID,  and it is show that number of rows is 104 that the same means that all vendor contains a product.

To re-check by second method, I check by normal query sum from BusinessEntityID join between 2 tables and search from a condition that if ProductID is null 

means no product as a result is zero.

7\. Create a SQL script that lookup for the customer in ‘Germany’ who did not use credit card.

Hint: Sales.Customer, Sales.SalesTerritory and Sales.PersonCreditCard

In [18]:
SELECT COUNT(Sales.PersonCreditCard.CreditCardID) AS No_CreditCard
FROM Sales.Customer
JOIN Sales.SalesTerritory ON Sales.Customer.TerritoryID = Sales.SalesTerritory.TerritoryID
JOIN Sales.PersonCreditCard ON Sales.PersonCreditCard.ModifiedDate = Sales.SalesTerritory.ModifiedDate
WHERE Sales.SalesTerritory.Name = 'Germany' AND Sales.PersonCreditCard.CreditCardID IS NULL

(1 row affected)

Total execution time: 00:00:00.207

No_CreditCard 0

8\. Create a SQL script that find the maximum or UnitPrice of each SalesOrderID and OrderDate

Hint: Sales.SalesOrderHeader and Sales.SalesOrderDetail

In [ ]:
SELECT Sales.SalesOrderHeader.SalesOrderID, Sales.SalesOrderHeader.OrderDate, MAX(Sales.SalesOrderDetail.UnitPrice) AS Max_UnitPrice
FROM Sales.SalesOrderHeader 
JOIN Sales.SalesOrderDetail ON Sales.SalesOrderDetail.SalesOrderID = Sales.SalesOrderHeader.SalesOrderID 
GROUP BY Sales.SalesOrderHeader.SalesOrderID, Sales.SalesOrderHeader.OrderDate


9\. Sales and Commission analysis: AdventureWork Limited need to analyze the sales and commission of the salesperson as the following criteria:

Output Column: 

Subcategory, Salesperson name (First name and Last name with space in between), 

NetSales (Must be calculated by UnitPrice x OrderQty) with the deducted tax of 3%, 

SalesCommission (Use NetSales with the following condition: 

(1) If the subcategory contains Bikes, the commission is 15% 

(2) If the subcategory is ‘Locks’ then no commission and 

(3) Other subcategory must be 10%

9.1. Create a SQL script from the condition above and must sort by Subcategory in ascending and SalesCommission from largest to least amount

In [21]:
SELECT Production.ProductSubcategory.Name AS Subcategory,
	   CONCAT(Person.Person.FirstName,' ',Person.Person.LastName) AS SalesPerson_Name, 
	   ((sod.UnitPrice * sod.OrderQty) - ((3 * (sod.UnitPrice * sod.OrderQty)) / 100)) AS NetSales,
	   CASE 
			WHEN Production.ProductSubcategory.Name LIKE '%Bikes%' THEN (15 * (sod.UnitPrice * sod.OrderQty)) / 100
			WHEN Production.ProductSubcategory.Name LIKE 'Locks' THEN (0 * (sod.UnitPrice * sod.OrderQty)) / 100
			ELSE (10 * (sod.UnitPrice * sod.OrderQty)) / 100
	   END AS SalesCommission
FROM Sales.SalesPerson 
JOIN Person.Person ON Person.Person.BusinessEntityID = Sales.SalesPerson.BusinessEntityID
JOIN Sales.SalesOrderHeader ON Sales.SalesOrderHeader.TerritoryID = Sales.SalesPerson.TerritoryID 
JOIN Sales.SalesOrderDetail sod ON sod.SalesOrderID = Sales.SalesOrderHeader.SalesOrderID
JOIN Production.Product ON Production.Product.ProductID = sod.ProductID 
JOIN Production.ProductSubcategory ON Production.Product.ProductSubcategoryID = Production.ProductSubcategory.ProductSubcategoryID
ORDER BY Subcategory ASC, SalesCommission DESC

Subcategory SalesPerson_Name NetSales SalesCommission Bib-Shorts Linda Mitchell 912.1837 94.0395 Bib-Shorts Shu Ito 912.1837 94.0395 Bib-Shorts Linda Mitchell 864.1740 89.0901 Bib-Shorts Shu Ito 864.1740 89.0901 Bib-Shorts Linda Mitchell 864.1740 89.0901 Bib-Shorts Shu Ito 864.1740 89.0901 Bib-Shorts Garrett Vargas 864.1740 89.0901 Bib-Shorts José Saraiva 864.1740 89.0901 Bib-Shorts Garrett Vargas 768.1547 79.1912 Bib-Shorts José Saraiva 768.1547 79.1912 Bib-Shorts Garrett Vargas 768.1547 79.1912 Bib-Shorts José Saraiva 768.1547 79.1912 Bib-Shorts Linda Mitchell 708.7973 73.0718 Bib-Shorts Shu Ito 708.7973 73.0718 Bib-Shorts Garrett Vargas 708.7973 73.0718 Bib-Shorts José Saraiva 708.7973 73.0718 Bib-Shorts Linda Mitchell 658.1689 67.8524 Bib-Shorts Shu Ito 658.1689 67.8524 Bib-Shorts Tsvi Reiter 607.5405 62.6330 Bib-Shorts Linda Mitchell 607.5405 62.6330 Bib-Shorts Shu Ito 607.5405 62.6330 Bib-Shorts Linda Mitchell 607.5405 62.6330 Bib-Shorts Shu Ito 607.5405 62.6330 Bib-Shorts Linda Mitchell 607.5405 62.6330 Bib-Shorts Shu Ito 607.5405 62.6330 Bib-Shorts Linda Mitchell 556.9122 57.4136 Bib-Shorts Shu Ito 556.9122 57.4136 Bib-Shorts Ranjit Varkey Chudukatil 556.9122 57.4136 Bib-Shorts Jae Pak 556.9122 57.4136 Bib-Shorts Michael Blythe 556.9122 57.4136 Bib-Shorts Garrett Vargas 556.9122 57.4136 Bib-Shorts José Saraiva 556.9122 57.4136 Bib-Shorts Pamela Ansman-Wolfe 556.9122 57.4136 Bib-Shorts David Campbell 556.9122 57.4136 Bib-Shorts Tete Mensa-Annan 556.9122 57.4136 Bib-Shorts Ranjit Varkey Chudukatil 556.9122 57.4136 Bib-Shorts Ranjit Varkey Chudukatil 556.9122 57.4136 Bib-Shorts Michael Blythe 523.7418 53.9940 Bib-Shorts Jillian Carson 523.7418 53.9940 Bib-Shorts Linda Mitchell 523.7418 53.9940 Bib-Shorts Shu Ito 523.7418 53.9940 Bib-Shorts Michael Blythe 523.7418 53.9940 Bib-Shorts Garrett Vargas 523.7418 53.9940 Bib-Shorts José Saraiva 523.7418 53.9940 Bib-Shorts Tsvi Reiter 523.7418 53.9940 Bib-Shorts Linda Mitchell 523.7418 53.9940 Bib-Shorts Shu Ito 523.7418 53.9940 Bib-Shorts Garrett Vargas 523.7418 53.9940 Bib-Shorts José Saraiva 523.7418 53.9940 Bib-Shorts Jae Pak 523.7418 53.9940 Bib-Shorts Linda Mitchell 523.7418 53.9940 Bib-Shorts Shu Ito 523.7418 53.9940 Bib-Shorts Linda Mitchell 471.3677 48.5946 Bib-Shorts Shu Ito 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Linda Mitchell 471.3677 48.5946 Bib-Shorts Shu Ito 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Linda Mitchell 471.3677 48.5946 Bib-Shorts Shu Ito 471.3677 48.5946 Bib-Shorts Linda Mitchell 471.3677 48.5946 Bib-Shorts Shu Ito 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Ranjit Varkey Chudukatil 471.3677 48.5946 Bib-Shorts Tsvi Reiter 471.3677 48.5946 Bib-Shorts Jillian Carson 471.3677 48.5946 Bib-Shorts Jillian Carson 471.3677 48.5946 Bib-Shorts Linda Mitchell 471.3677 48.5946 Bib-Shorts Shu Ito 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Jillian Carson 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 471.3677 48.5946 Bib-Shorts José Saraiva 471.3677 48.5946 Bib-Shorts Garrett Vargas 418.9935 43.1952 Bib-Shorts José Saraiva 418.9935 43.1952 Bib-Shorts Ranjit Varkey Chudukatil 418.9935 43.1952 Bib-Shorts Linda Mitchell 418.9935 43.1952 Bib-Shorts Shu Ito 418.9935 43.1952 Bib-Shorts Garrett Vargas 418.9935 43.1952 Bib-Shorts José Saraiva 418.9935 43.1952 Bib-Shorts Linda Mitchell 418.9935

9.2. Create a summary from above with CTE or Derived table by sum only sales commission of each salesperson with the TotalCommission greater than 25,000 

and sort by TotalCommission in descending

In [ ]:
;WITH cte 
AS
(
	SELECT Production.ProductSubcategory.Name AS Subcategory,
		   CONCAT(Person.Person.FirstName,' ',Person.Person.LastName) AS SalesPerson_Name, 
	       ((sod.UnitPrice * sod.OrderQty) - ((3 * (sod.UnitPrice * sod.OrderQty)) / 100)) AS NetSales,
		   CASE 
				WHEN Production.ProductSubcategory.Name LIKE '%Bikes%' THEN (15 * (sod.UnitPrice * sod.OrderQty)) / 100
				WHEN Production.ProductSubcategory.Name LIKE 'Locks' THEN (0 * (sod.UnitPrice * sod.OrderQty)) / 100
				ELSE (10 * (sod.UnitPrice * sod.OrderQty)) / 100
			END AS SalesCommission
	FROM Sales.SalesPerson 
	JOIN Person.Person ON Person.Person.BusinessEntityID = Sales.SalesPerson.BusinessEntityID
	JOIN Sales.SalesOrderHeader ON Sales.SalesOrderHeader.TerritoryID = Sales.SalesPerson.TerritoryID 
	JOIN Sales.SalesOrderDetail sod ON sod.SalesOrderID = Sales.SalesOrderHeader.SalesOrderID
	JOIN Production.Product ON Production.Product.ProductID = sod.ProductID 
	JOIN Production.ProductSubcategory ON Production.Product.ProductSubcategoryID = Production.ProductSubcategory.ProductSubcategoryID
),
cte2 
AS
(
	SELECT SalesPerson_Name, Sum(SalesCommission) AS TotalCommission 
	FROM cte
	GROUP BY SalesPerson_Name
)

SELECT * FROM cte2
WHERE TotalCommission > 25000
ORDER BY TotalCommission DESC